### DISCUSSION TITLE
Problems with r_in parameter for atpass()

### DISCUSSION THREAD

**Discussion by tfs563** [atcollab/at#634] [Python AT]
Edit: I now have a similar question that follows under this title. 

I'm trying a more simplistic particle tracking exercise by creating a single array to be used as the r_in parameter of the at.pass() function, but I keep getting errors such as 

ValueError: rin is not a double array

However, when I check the shape of the array I made, I get exactly the type of array outlined in the parameters breakdown for r_in, which says it must be a (6, N) array. 

In my code, I first defined the lattice as 
```python
x = -1

Dr1 = at.Drift('Dr', 0.08)
Dr2 = at.Drift('Dr', 0.02)
Dr3 = at.Drift('Dr', 0.054)
Dr4 = at.Drift('Dr', 0.066)
QF1 = at.Quadrupole('QF', 0.025, x * 3.2)
QF2 = at.Quadrupole('QF', 0.04, 4)
QF3 = at.Quadrupole('QF', 0.025, x * 3.2)
QD1 = at.Quadrupole('QD', 0.025, x * -3.2)
QD2 = at.Quadrupole('QD', 0.045, x * -3.2)
Bend1 = at.Dipole('Bend', 0.15, 2 * pi / 6)

# Defines the lattice of the entire energy filter
Energy_Filter = at.Lattice([Dr1, QD1, Dr2, QF1, Dr3, Bend1, 
                            Dr4, QF2, Dr4, Bend1, Dr3, QF3, Dr2, QD2, Dr1], 
                           name='Energy_Filter_MAX_IV', energy=2E3)
``` 
and then create the array while passing it into the atpass() function using 
```python
rin1 = np.array([0, 0, 0, 0, 0, 0]) # initial distribution of a particle
rin = np.asfortranarray(np.vstack(rin1))
print(rin.shape) 

lattice_range = range(len(Energy_Filter)+1)
refpts = np.ndarray(lattice_range, np.dtype(np.uint32))

# Tracks the particle through the lattice
result1 = at.atpass(Energy_Filter, rin, 1, refpts)
print(result1)
``` 

printing the shape of rin gives me a (6, 1) array, but I still get the value error. So, what is it that I'm doing wrong to keep getting this error?

**Comment by swhite2401** [comment]
Hello,
I have several comments on your example.
First the error you get is "ValueError: rin is not a double array", this points to a typing error: you have to provide a double array.

You may change:
`rin1 = np.array([0, 0, 0, 0, 0, 0]) # initial distribution of a particle`
to
`rin1 = np.zeros(6) # initial distribution of a particle`

Then at atpass will give you an error because it is a private interface and will point you to at.lattice_pass that is the official public tracking interface.
The help for this function is there: https://atcollab.github.io/at/p/api/at.tracking.track.html#at.tracking.track.lattice_pass
 and the syntax is:
`at.lattice_pass(ring, rin, nturns=1)` or `ring.lattice_pass(rin, nturns=1)`

**Reply by tfs563** [reply]
This works and gives me a result, thanks!
